test/k8sT/Services.go

```diff
@@ -721,6 +723,107 @@ var _ = Describe("K8sServicesTest", func() {
 			kubectl.CiliumExecMustSucceed(context.TODO(), pod, "cilium bpf ct flush global", "Unable to flush CT maps")
 		}
 
+		testSessionAffinity := func() {
+			var (
+				data   v1.Service
+				dstPod string
+				count  = 10
+			)
+
+			err := kubectl.Get(helpers.DefaultNamespace, "service test-affinity").Unmarshal(&data)
+			Expect(err).Should(BeNil(), "Cannot retrieve service")
+			externalNodeName, _ := kubectl.GetNodeInfo(helpers.GetNodeWithoutCilium())
+			_, k8s1IP := kubectl.GetNodeInfo(helpers.K8s1)
+
+			httpURL := getHTTPLink(k8s1IP, data.Spec.Ports[0].NodePort)
+			cmd := helpers.CurlFail(httpURL) + " | grep 'Hostname:' " // pod name is in the hostname
+
+			// Send 10 requests to he test-affinity NodePort svc from outside (lb.h),
+			// and check that the same backend is chosen (checkDstPodIsSame=true)
+
+			By("Making %d HTTP requests from outside cluster to %q (sessionAffinity)", count, httpURL)
+
+			for i := 1; i <= count; i++ {
+				res, err := kubectl.ExecInHostNetNS(context.TODO(), externalNodeName, cmd)
+				Expect(err).Should(BeNil(), "Cannot exec in %s host netns", externalNodeName)
+				ExpectWithOffset(1, res).Should(helpers.CMDSuccess(),
+					"Cannot connect to service %q from outside cluster (%d/%d)", httpURL, i, count)
+				pod := strings.TrimSpace(strings.Split(res.GetStdOut(), ": ")[1])
+				if i == 1 {
+					// Retrieve the destination pod from the first request
+					dstPod = pod
+				} else {
+					// Check that destination pod is always the same
+					Expect(dstPod).To(Equal(pod))
+				}
+			}
+
+			// Delete the pod, and check that a new backend is chosen
+			kubectl.DeleteResource("pod", dstPod).ExpectSuccess("Unable to delete %s pod", dstPod)
+			kubectl.WaitPodDeleted(dstPod)
+			// Unfortunately, it takes a while until cilium-agent receives Endpoint
+			// update event which triggers a removal of the deleted dstPod from
+			// the affinity and the service BPF maps. Therefore, the bellow requests
+			// are flaky.
+			// TODO(brb) don't sleep, instead wait for Endpoint obj update (might be complicated though)
```

Comment:
Is this TODO still relevant? Is it meant to be implemented for this PR?

Reply:
Yep, it's still relevant. I will address it in a follow-up PR.

Reply:
Maybe use `EndpointGet()` helper to wait the Cilium Endpoint is deleted?

Reply:
Hmm, but that might be not enough, as we want to make sure that cilium-agent has processed the related Endpoint object removal event, while `EndpointGet()` gets updated via CNI DEL (happens before the former).